### PR TITLE
Resize channel buffers

### PR DESCRIPTION
**Pull request by gaaclarke**
Implemented a mechanism for BasicMessageChannels to resize their Channel Buffer.

Under certain circumstances sending messages on channels can result in messages getting thrown away.  Channel Buffers added the ability for the engine to store those messages until they can be delivered.  Now you can resize those buffers to adjust the behavior of the channel - do you want to store just the latest version? Nothing? Or attempt to store everything?

Related issue: flutter/flutter#38914

### CODE REVIEW

File: lib/ui/channel_buffers.dart [hunk 1]
```diff
@@ -112,6 +112,8 @@ class ChannelBuffers {
   /// buffer size that will avoid any overflows.
   static const int kDefaultBufferSize = 1;
 
+  static const String kControlChannelName = "com.google.flutter/channel-buffers";
```
Comment (xster) [inline lib/ui/channel_buffers.dart]: dev.flutter/

Comment (gaaclarke) [reply lib/ui/channel_buffers.dart]: Done

File: lib/ui/channel_buffers.dart [hunk 2]
```diff
@@ -185,6 +187,28 @@ class ChannelBuffers {
...
     }
   }
+
+  String _getString(ByteData data) {
+    final ByteBuffer buffer = data.buffer;
+    final Uint8List list = buffer.asUint8List(data.offsetInBytes, data.lengthInBytes);
+    return utf8.decode(list);
+  }
+
+  /// Send a control message.
```
Comment (xster) [inline lib/ui/channel_buffers.dart]: Somewhat unexpected that the name here is send a control message. 

Wouldn't the verb be intercept or monitor?

File: lib/ui/channel_buffers.dart [hunk 3]
```diff
@@ -185,6 +187,28 @@ class ChannelBuffers {
...
+  ///
+  /// This is intended to be called by the platform messages dispatcher.
+  ///
+  /// Available messages:
+  /// * resize - Allows you to set the size of a channel's buffer, command is in the format of
+  ///  `resize/<channel name>/<new size>`
+  void handleMessage(ByteData data) {
+    final List<String> command = _getString(data).split('\r');
+    if (command.length == 3 && command[0] == 'resize') {
```
Comment (xster) [inline lib/ui/channel_buffers.dart]: Add code comment for the 3

File: lib/ui/hooks.dart [hunk 4]
```diff
@@ -150,7 +150,15 @@ void _updateAccessibilityFeatures(int values) {
 
 @pragma('vm:entry-point')
 void _dispatchPlatformMessage(String name, ByteData data, int responseId) {
-  if (window.onPlatformMessage != null) {
+  if (name == ChannelBuffers.kControlChannelName) {
+    try {
+      channelBuffers.handleMessage(data);
+    } catch (ex) {
+      _printDebug('Message "$name" caused exception $ex');
```
Comment (xster) [inline lib/ui/hooks.dart]: This here would say

"Message dev.flutter/channel-buffers caused exception... " right? Maybe Message to dev.flutter/channel-buffers?

File: lib/ui/channel_buffers.dart (hunk 3 above)
Comment (gaaclarke) [reply lib/ui/channel_buffers.dart]: done

File: lib/ui/hooks.dart (hunk 4 above)
Comment (gaaclarke) [reply lib/ui/hooks.dart]: done

File: lib/ui/channel_buffers.dart (hunk 2 above)
Comment (gaaclarke) [reply lib/ui/channel_buffers.dart]: switched to "handle"

File: lib/ui/channel_buffers.dart [hunk 5]
```diff
@@ -185,6 +187,30 @@ class ChannelBuffers {
...
+  /// Handle a control message.
+  ///
+  /// This is intended to be called by the platform messages dispatcher.
+  ///
+  /// Available messages:
+  /// - Name: resize
+  ///   Arity: 2
+  ///   Format: `resize\r<channel name>\r<new size>`
+  ///   Description: Allows you to set the size of a channel's buffer.
+  void handleMessage(ByteData data) {
```
Comment (chinmaygarde) [inline lib/ui/channel_buffers.dart]: Is the structure of these control messages, name of the channel, etc. documented anywhere or is this a new convention?

File: lib/ui/channel_buffers.dart [hunk 6]
```diff
@@ -162,7 +164,7 @@ class ChannelBuffers {
   ///
   /// This could result in the dropping of messages if newSize is less
   /// than the current length of the queue.
-  void resize(String channel, int newSize) {
+  void _resize(String channel, int newSize) {
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply lib/ui/channel_buffers.dart]: It was public only for about a week.

File: lib/ui/channel_buffers.dart (hunk 5 above)
Comment (gaaclarke) [reply lib/ui/channel_buffers.dart]: It's new.  I've documented them in the docstring.  If there is somewhere else you'd like to see them let me know.  The intention is that there is nice API's overtop of them so no one has to create them manually.

**Pull request merged**